web.config

```
@@ -228,41 +228,66 @@
+                  <match url="^((.+)/)?index(\.php)?$" />
+                  <action type="Redirect" url="{R:1}" />
+                </rule>
+
+                <rule name="Remove .php extension and redirect" stopProcessing="true">
                   <match url="^(.+)\.php$" />
                   <conditions logicalGrouping="MatchAll" trackAllCaptures="false">
                     <add matchType="IsFile" />
                   </conditions>
                   <action type="Redirect" url="{R:1}" />
                 </rule>
 
-                <rule name="Redirect index or index.php" stopProcessing="true">
-                  <match url="((.+)/)?index(\.php)?$" />
-                  <action type="Redirect" url="{R:1}" />
+                <rule name="Redirect folder without / to include /" stopProcessing="true">
+                  <match url="^(.+[^/])$" />
+                  <conditions logicalGrouping="MatchAll" trackAllCaptures="false">
+                    <add matchType="IsDirectory" />
+                    <add matchType="IsFile" negate="true" input="{DOCUMENT_ROOT}/{R:1}.php" />
+                    <add matchType="IsFile" negate="true" input="{DOCUMENT_ROOT}/{R:1}.md" />
+                  </conditions>
+                  <action type="Redirect" url="{R:1}/" />
                 </rule>
 
-                <rule name="Ignore folders" stopProcessing="true">
-                  <match url="^(.+)/$" />
-                  <conditions logicalGrouping="MatchAll" trackAllCaptures="false" />
-                  <action type="Rewrite" url="{R:1}/index.php" />
-                </rule>
+                <!-- PHP rewriting -->
 
-                <rule name="Is a folder" stopProcessing="true">
+                <rule name="Rewrite file to file.md" stopProcessing="false">
                   <match url="^(.+)$" />
                   <conditions logicalGrouping="MatchAll" trackAllCaptures="false">
-                    <add matchType="IsDirectory" />
+                    <add matchType="IsFile" input="{DOCUMENT_ROOT}/{R:1}.md" />
                   </conditions>
-                  <action type="Redirect" url="{R:1}/" />
+                  <action type="Rewrite" url="/_common/mdhost.php?file={R:1}.md" />
                 </rule>
 
-                <rule name="Add .php extension">
+                <rule name="Rewrite file to file.php" stopProcessing="false">
                   <match url="^(.+)$" />
                   <conditions logicalGrouping="MatchAll" trackAllCaptures="false">
-                    <add matchType="IsFile" negate="True" />
+                    <add matchType="IsFile" input="{DOCUMENT_ROOT}/{R:1}.php" />
+                    <add matchType="IsFile" negate="true" input="{DOCUMENT_ROOT}/{R:1}.md" />
                   </conditions>
                   <action type="Rewrite" url="{R:1}.php" />
                 </rule>
+
+                <rule name="Rewrite folder/ to folder/index.md" stopProcessing="false">
```

```suggestion
                <rule name="Rewrite folder/ to folder/index.md" stopProcessing="true">
```

This should be a final action.